src/Parsing/String/Replace.purs

```diff
@@ -0,0 +1,545 @@
+-- |   string-to-number conversion function fails. A typed parser will perform both
+-- |   the pattern match and the conversion, so it will never be in that situation.
+-- |   [Parse, don't validate.](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)
+-- |
+-- | * Regular expressions are only able to pattern-match
+-- |   [regular grammars](https://en.wikipedia.org/wiki/Chomsky_hierarchy#The_hierarchy).
+-- |   Monadic parsers are able pattern-match context-free (by recursion)
+-- |   or context-sensitive (by monad transformer) grammars.
+-- |
+-- | * The replacement expression for a traditional regular expression-based
+-- |   substitution command is usually just a string template in which
+-- |   the *Nth* “capture group” can be inserted with the syntax `\N`. With
+-- |   this library, instead of a template, we get
+-- |   an `editor` function which can perform any computation, including `Effect`s.
+-- |
+-- | #### Implementation Notes
+-- |
+-- | All of the functions in this module work by calling `runParserT`
+-- | with the `anyTill` combinator.
+-- | We can expect the speed of parser-based pattern matching to be
+-- | about 10× worse than regex-based pattern matching in a JavaScript
+-- | runtime environment.
+-- | This module is based on the Haskell packages
+-- | [__replace-megaparsec__](https://hackage.haskell.org/package/replace-megaparsec)
+-- | and
+-- | [__replace-attoparsec__](https://hackage.haskell.org/package/replace-attoparsec).
+-- |
+module Parsing.String.Replace
+  ( breakCap
+  , breakCapT
+  , splitCap
+  , splitCapT
+  , streamEdit
+  , streamEditT
+  ) where
+
+import Prelude
+
+import Control.Monad.Rec.Class (class MonadRec, Step(..), tailRecM)
+import Control.Monad.ST (ST)
+import Control.Monad.ST.Global (Global, toEffect)
+import Data.Array as Array
+import Data.Array.ST (STArray)
+import Data.Array.ST as Array.ST
+import Data.Array.ST.Partial as Array.ST.Partial
+import Data.Either (Either(..), hush)
+import Data.List (List(..), foldl, uncons, (:))
+import Data.List as List
+import Data.List.NonEmpty (NonEmptyList, fromList, singleton)
+import Data.Maybe (Maybe(..), fromJust, isJust, maybe)
+import Data.Newtype (unwrap, wrap)
+import Data.Nullable (Nullable, notNull, null)
+import Data.String (CodePoint, joinWith)
+import Data.String as CodePoint
+import Data.String as String
+import Data.Tuple (Tuple(..))
+import Data.Tuple.Nested (T3, (/\))
+import Effect.Unsafe (unsafePerformEffect)
+import Parsing (ParseState(..), Parser, ParserT, Position(..), getParserT, position, runParserT)
+import Parsing.Combinators (optionMaybe, try)
+import Parsing.String (anyCodePoint, anyTill, rest)
+import Partial.Unsafe (unsafePartial)
+import Unsafe.Coerce (unsafeCoerce)
+
+-- | Monad transformer version of `breakCap`. The `sep` parser will run
+-- | in the monad context.
+breakCapT
+  :: forall m a
+   . (Monad m)
+  => (MonadRec m)
+  => String
+  -> ParserT String m a
+  -> m (Maybe (T3 String a String))
```

Comment:
Would a record be more appropriate rather than T3?

Reply:
Maybe so? Like if this returned `m (Maybe ({ prefix :: String, cap :: a, suffix :: String}))`?

I kind of like the positional `prefix /\ cap /\ suffix` syntax because the relationship between the three return values is positional.